feat: BROS-199: Enhanced paragraph annotation with auto-selection

cloudmark opened the pull request:
## **BROS-199: Enhanced Paragraph Selection with "Select All" Functionality**

### **Reason for Change**

This PR implements an enhanced paragraph selection feature that allows users to efficiently select and annotate multiple phrases in paragraph-based annotation tasks. The implementation includes:

- **One-click phrase selection**: Click on any phrase to select it and automatically apply selected labels
- **Proactive text selection handling**: Preserves user text selections for more precise annotation
- **Enhanced keyboard shortcuts**: Added hotkeys for improved annotation workflow
- **Improved visual feedback**: Better styling and highlighting for selected phrases
- **Robust reaction management**: Proper cleanup of MobX reactions to prevent memory leaks

### **Key Features**

1. **Smart Phrase Selection**: Clicking on a phrase automatically selects it and applies any currently selected labels
2. **Text Selection Preservation**: User text selections are proactively saved and used for more precise annotations
3. **Automatic Region Management**: Unselects regions that don't belong to the current phrase context
4. **Keyboard Shortcuts**: Added configurable hotkeys for faster annotation workflows
5. **Feature Flag Controlled**: All functionality is gated behind the `FF_NER_SELECT_ALL` feature flag

### **Technical Changes**

#### **Frontend Changes**
- **`HtxParagraphs.jsx`**: Major enhancements (+307 lines) including phrase selection logic, text selection preservation, and improved event handling
- **`Phrases.jsx`**: Updated phrase rendering with enhanced selection capabilities (+123 lines)
- **`model.js`**: Added MobX reactions for label selection changes, with proper cleanup mechanisms (+40 lines)
- **`Paragraphs.module.scss`**: Enhanced styling for better visual feedback (+109 lines)

#### **Configuration Updates**
- **`keymap.json`**: Added new keyboard shortcuts for annotation workflows
- **`defaults.js`**: Updated hotkey configurations for account settings
- **`feature-flags.ts`**: Added `FF_NER_SELECT_ALL` feature flag definition

#### **Testing**
- **`Phrases.test.tsx`**: Comprehensive test coverage (+373 lines) for all new functionality
- **`AtParagraphs.js`**: Added E2E test fragments for automated testing

### **Rollout Strategy**

- **Feature Flag**: All functionality is controlled by the `FF_NER_SELECT_ALL` feature flag
- **Gradual Rollout**: Feature can be enabled for specific users/organizations initially
- **Backward Compatibility**: All existing functionality remains unchanged when feature flag is disabled

### **Testing**

✅ **Unit Tests**: All new functionality covered with comprehensive test suite
✅ **E2E Tests**: Updated automation fragments for end-to-end testing
✅ **Manual Testing**: Verified functionality across different annotation scenarios
✅ **Performance Testing**: MobX reaction cleanup prevents memory leaks
✅ **Regression Testing**: Existing paragraph functionality remains unaffected

### **Screenshots**

*Loom Video Reference* : https://www.loom.com/share/87a82eb4a4c7469fa53440c50eaec681


### **Reviewer Notes**

- **Feature Flag Usage**: All new functionality is properly gated behind `FF_NER_SELECT_ALL`

### **General Notes**

This implementation significantly improves the annotation experience for paragraph-based tasks, particularly for NER (Named Entity Recognition) workflows. The proactive text selection handling and automatic phrase annotation features reduce the number of clicks required for efficient labeling while maintaining precision and user control.

The feature is designed to be non-intrusive and can be safely rolled out gradually through the feature flag system.

robot-ci-heartex committed:
ci: Update Feature Flags

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16489122214

hlomzik commented on web/libs/editor/src/regions/ParagraphsRegion.js:
```diff
@@ -91,6 +91,10 @@ const Model = types
       return res;
     },
+
+    onClickRegion() {
```
do we need this? it should be inherited from `RegionsMixin`

hlomzik commented on web/libs/editor/src/tags/object/Paragraphs/HtxParagraphs.jsx:
```diff
@@ -518,19 +613,216 @@ class HtxParagraphsView extends Component {
+    if (!item || !item.annotation || !item.annotation.results) return;
+
+    // Only deselect regions, not labels
+    this.unselectAllRegions();
```
we already have this method
```suggestion
    item.annotation.unselectAreas();
```

hlomzik commented on web/libs/editor/src/tags/object/Paragraphs/HtxParagraphs.jsx:
```diff
@@ -518,19 +613,216 @@ class HtxParagraphsView extends Component {
+
+    const item = this.props.item;
+
+    if (item.annotation.selectArea && typeof item.annotation.selectArea === "function") {
```
What might go wrong here? I'd suggest to remove this method and use annotation method directly, possibly with optional chaining

hlomzik commented on web/libs/editor/src/tags/object/Paragraphs/HtxParagraphs.jsx:
```diff
@@ -518,19 +613,216 @@ class HtxParagraphsView extends Component {
+  };
+
+  // Get all regions for the current phrase
+  getRegionsForPhrase = (phraseIdx) => {
```
I think this method belongs to the model

hlomzik commented on web/libs/editor/src/tags/object/Paragraphs/HtxParagraphs.jsx:
```diff
@@ -518,19 +613,216 @@ class HtxParagraphsView extends Component {
+    const item = this.props.item;
+    if (!item || !item._value || item._value.length === 0) return;
+    const idx = typeof item.playingId === "number" && item.playingId >= 0 ? item.playingId : 0;
+    if (item.selectAndAnnotatePhrase) {
```
Why won't it be there? Which scenario will lead us there?

asynclizard committed:
fix: move phrase navigation and annotation logic to model, clean up view

- Move business logic for phrase navigation (next/previous) and "select all and annotate" from HtxParagraphs view to Paragraphs model as new model actions:
    - goToNextPhrase
    - goToPreviousPhrase
    - selectAllAndAnnotateCurrentPhrase
- Update view to call new model methods, simplifying UI code and improving separation of concerns
- Remove over-defensive type checks for playingId and method existence; use direct calls and Math.max for index safety
- Remove redundant region/label unselection helpers in favor of built-in annotation methods
- Ensure all navigation and annotation features are test-verified and work in both audio and no-audio scenarios

This refactor improves maintainability, testability, and code clarity by centralizing business logic in the model and keeping the view focused on UI.

hlomzik commented on web/libs/editor/src/tags/object/Paragraphs/model.js:
```diff
@@ -412,6 +413,86 @@ const PlayableAndSyncable = types
+
+      if (existingRegions.length > 0) {
+        // Clear any existing selections and return early - no text selection, no annotation
+        if (self.annotation && self.annotation.unselectAll) {
+          self.annotation.unselectAll();
+        }
+        return;
+      }
+
+      // Only select text and create annotation if no duplicates found
+      self.selectPhraseText(phraseIndex);
+      if (self._viewRef && typeof self._viewRef.createAnnotationForPhrase === "function") {
+        self._viewRef.createAnnotationForPhrase(phraseIndex);
+      }
```
Thank you for changes! I would also simplify things like this:
- `self.annotation?.unselectAll?.()` — and we wouldn't have annotation at this point?
- `self._viewRef?.createAnnotationForPhrase(phraseIndex)` — if ref is set we know its type so function will be there for sure

niklub commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/16579834313)
> Successfully merged:  create mode 100644 web/libs/ui/src/assets/icons/cloud-provider-s3.svg

niklub commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/16588090690)
> Successfully merged:  create mode 100644 web/libs/ui/src/assets/icons/cloud-provider-s3.svg

niklub committed:
Merge branch 'develop' into 'fb-bros-199-select-all-variant'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16588090690

asynclizard committed:
fix: modal is failing on the test e2e, moving to Modal.seeWarning instead

hlomzik commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/16593924689)
> Successfully merged: Already up to date.

asynclizard committed:
fix: different approach to testing the paragraphs (so we avoid the modal)

hlomzik commented on web/libs/editor/src/tags/object/Paragraphs/Paragraphs.module.scss:
```diff
@@ -7,12 +7,35 @@ $border-thin: 1px solid rgb(137 128 152 / 16%);
     content: counter(phrase);
     margin-right: 4px;
   }
+
+  // Only add padding when select all button is present
+  &.withSelectAllButton {
+    padding-right: 50px !important; // Override any other padding from newUI
+  }
+  
+  // Ensure newUI doesn't add extra padding unless select all button is present
+  &.newUI {
+    // Don't add any extra padding by default
+    &:not(.withSelectAllButton) {
+      padding-right: 0.5em !important; // Reset to default, override any inherited padding
+    }
+    
+    &.withSelectAllButton {
+      padding-right: 50px !important; // Only add extra padding when button is present
+    }
+  }
+}
```
this all seems a little excess, you add this padding later in NewUI and it looks correct

hlomzik commented on web/libs/editor/src/tags/object/Paragraphs/Phrases.jsx:
```diff
@@ -177,35 +195,77 @@ export const Phrases = observer(({ item, playingId, activeRef, setIsInViewport }
+          }}
+          tabIndex={idx}
+          data-testid={`phrase:${idx}`}
+          className={`${classNames.join(" ")} ${isFF(FF_NER_SELECT_ALL) && isActive ? styles.activePhrase : ""}`}
```
second part can also be moved to the code above with `if`s for consistency and readability

asynclizard committed:
fix: improve phrase timing and hotkey navigation in Paragraphs

- Add audio duration fallback for phrases without explicit end times
- Fix hotkey navigation focus for testing
- Remove redundant CSS padding logic
- Fix "Show region labels" setting refresh issue
- Fix "jump" glitch

robot-ci-heartex committed:
ci: Build tag docs

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16621332018

asynclizard committed:
fix: make tests more stable

asynclizard committed:
fix: the span visibility was also incorrect in the HighlightMixin

niklub commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/16672272417)
> Successfully merged:  delete mode 100644 web/libs/editor/src/components/Timeline/Controls/VideoConfigControl.scss

niklub committed:
Merge branch 'develop' into 'fb-bros-199-select-all-variant'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16672272417

niklub commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/16690871062)
> Successfully merged:  18 files changed, 207 insertions(+), 81 deletions(-)

niklub committed:
Merge branch 'develop' into 'fb-bros-199-select-all-variant'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16690871062
